src/app/renderer/canvas.tsx

```diff
@@ -1,22 +1,105 @@
 import { Graphics, Rect, Color, Point, prettyNumber } from "../../core";
+import { MarkElement } from "../../core/graphics";
 
-function renderColor(color: Color, opacity?: number): string {
+// adapted from https://stackoverflow.com/a/20820649
+function desaturate(color: Color, amount: number) {
+  const { r, g, b } = color;
+  const l = 0.3 * r + 0.6 * g + 0.1 * b;
+  return {
+    r: Math.min(r + amount * (l - r), 255),
+    g: Math.min(g + amount * (l - g), 255),
+    b: Math.min(b + amount * (l - b), 255)
+  };
+}
+
+function renderColor(
+  color: Color,
+  opacity?: number,
+  applyDesaturation = false
+): string {
+  if (applyDesaturation) {
+    color = desaturate(color, 0.7);
+  }
   return `rgba(${color.r.toFixed(0)},${color.g.toFixed(0)},${color.b.toFixed(
     0
   )},${opacity != undefined ? prettyNumber(opacity) : 1})`;
 }
 
+export interface CanvasRenderOptions {
+  width: number;
+  height: number;
+  hitTest: boolean;
+  selectedIndexes: number[];
+}
+
+interface RenderState {
+  options: CanvasRenderOptions;
+  hit?: {
+    colorToElement: Map<string, Graphics.Element>;
+    elementToColor: Map<Graphics.Element, string>;
+  };
+}
+
+/**
+ * Returns a unique color for the given element to be used when rendering the hit target canvas
+ * @param renderState The current render state
+ * @param element The element to get the hit testing color for
+ */
+function getHitColor(
+  renderState: RenderState,
+  element: Graphics.Element
+): string {
+  const { elementToColor, colorToElement } = renderState.hit;
+  let color = elementToColor.get(element);
+  if (!color) {
+    // Preferring random here, cause it helps a lot with the false positives from anti-aliasing
```

Comment:
We can disable anti-aliasing: https://stackoverflow.com/questions/195262/can-i-turn-off-antialiasing-on-an-html-canvas-element

Reply:
I saw that, but it looks like it only applies to images, did I miss something?  Or are you talking about the translation by .5 pixel?  I was reading somewhere else that it helps a bit, but it doesn't fully solve the problem.

Reply:
Oh, you are right, I didn't notice it's for images.

Reply:
The .5 pixel translation won't work since we are using a lot of non-integer coordinates and group transforms.